File: code/data_32.cs
using System;
using System.Text;

public static class Base32
{
    private const string Base32Alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567";


    public static string Encode(byte[] data)
    {
        if (data == null || data.Length == 0)
        {
            return string.Empty;
        }

        StringBuilder result = new StringBuilder();
        int bitsInBuffer = 0;
        int buffer = 0;

        foreach (byte b in data)
        {
            buffer = (buffer << 8) | b;
            bitsInBuffer += 8;

            while (bitsInBuffer >= 5)
            {
                bitsInBuffer -= 5;
                result.Append(Base32Alphabet[(buffer >> bitsInBuffer) & 0x1F]);
            }
        }

        if (bitsInBuffer > 0)
        {
            result.Append(Base32Alphabet[(buffer << (5 - bitsInBuffer)) & 0x1F]);
        }

        // Add padding if necessary (RFC 3548 compliant, but often omitted in practice)
        while (result.Length % 8 != 0)
        {
            result.Append('=');
        }

        return result.ToString();
    }

    public static byte[] Decode(string base32String)
    {
        if (string.IsNullOrEmpty(base32String))
        {
            return new byte[0];
        }

        base32String = base32String.TrimEnd('='); // Remove padding characters

        int bitsInBuffer = 0;
        int buffer = 0;
        byte[] decodedBytes = new byte[base32String.Length * 5 / 8]; // Max possible size
        int byteIndex = 0;

        foreach (char c in base32String)
        {
            int charValue = Base32Alphabet.IndexOf(char.ToUpper(c));
            if (charValue == -1)
            {
                throw new ArgumentException("Invalid Base32 character encountered.");
            }

            buffer = (buffer << 5) | charValue;
            bitsInBuffer += 5;

            if (bitsInBuffer >= 8)
            {
                bitsInBuffer -= 8;
                decodedBytes[byteIndex++] = (byte)(buffer >> bitsInBuffer);
            }
        }

        Array.Resize(ref decodedBytes, byteIndex); // Resize to actual length
        return decodedBytes;
    }
}

File: code/data32.java
public class Base32Codec {

    private static final String BASE32_ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567";

    public static String encodeBase32(byte[] data) {
        StringBuilder binaryString = new StringBuilder();
        for (byte b : data) {
            
            binaryString.append(String.format("%8s", Integer.toBinaryString(b & 0xFF)).replace(' ', '0'));
        }

        int paddingBits = 5 - (binaryString.length() % 5);
        if (paddingBits != 5) { 
            binaryString.append("0".repeat(paddingBits));
        }

        StringBuilder encoded = new StringBuilder();
        for (int i = 0; i < binaryString.length(); i += 5) {
            String fiveBits = binaryString.substring(i, i + 5);
            int index = Integer.parseInt(fiveBits, 2);
            encoded.append(BASE32_ALPHABET.charAt(index));
        }

        
        while (encoded.length() % 8 != 0) {
            encoded.append('=');
        }

        return encoded.toString();
    }

    
    public static byte[] decodeBase32(String data) {
        
        data = data.replace("=", "");

        StringBuilder binaryString = new StringBuilder();
        for (char c : data.toCharArray()) {
            int index = BASE32_ALPHABET.indexOf(c);
            if (index == -1) {
                throw new IllegalArgumentException("Invalid Base32 character: " + c);
            }
           
            binaryString.append(String.format("%5s", Integer.toBinaryString(index)).replace(' ', '0'));
        }

        
        int paddingBits = binaryString.length() % 8;
        if (paddingBits != 0) {
            binaryString.setLength(binaryString.length() - paddingBits);
        }

        byte[] decoded = new byte[binaryString.length() / 8];
        for (int i = 0; i < binaryString.length(); i += 8) {
            String eightBits = binaryString.substring(i, i + 8);
            decoded[i / 8] = (byte) Integer.parseInt(eightBits, 2);
        }

        return decoded;
    }

    public static void main(String[] args) {
        String originalText = "Hello, Base32!";
        byte[] originalBytes = originalText.getBytes();
    /*Examples (REMOVE IF NOT WANTED)
        // Encode
        String encodedText = encodeBase32(originalBytes);
        System.out.println("Original Text: " + originalText);
        System.out.println("Encoded Base32: " + encodedText);

        // Decode
        byte[] decodedBytes = decodeBase32(encodedText);
        String decodedText = new String(decodedBytes);
        System.out.println("Decoded Text: " + decodedText); */
    }
}
